orchagent/tunneldecaporch.cpp

```diff
@@ -0,0 +1,195 @@
+                }
+                if (fvField(i) == "dscp_mode")
+                {
+                    dscp_mode = fvValue(i);
+                }
+                if (fvField(i) == "ecn_mode")
+                {
+                    ecn_mode = fvValue(i);
+                }
+                if (fvField(i) == "ttl_mode")
+                {
+                    ttl_mode = fvValue(i);
+                }
+            }
+            if (addDecapTunnel(tunnel_type, ip_addresses, dscp_mode, ecn_mode, ttl_mode)) 
+            {
+                SWSS_LOG_NOTICE("Tunnel(s) added to ASIC_DB.");
+            }
+            else
+            {
+                SWSS_LOG_ERROR("Failed to add tunnels to ASIC_DB.");
+            }
+        }
+
+        it = consumer.m_toSync.erase(it);
+    }
+}
+
+/**
+ * Function Description:
+ *    @brief adds a decap tunnel entry to ASIC_DB
+ *
+ * Arguments:
+ *    @param[in] type - type of tunnel
+ *    @param[in] dst_ip - destination ip address to decap
+ *    @param[in] dscp - dscp mode (uniform/pipe)
+ *    @param[in] ecn - ecn mode (copy_from_outer/standard)
+ *    @param[in] ttl - ttl mode (uniform/pipe)
+ *
+ * Return Values:
+ *    @return true on success and false if there's an error
+ */
+bool TunnelDecapOrch::addDecapTunnel(string type, IpAddresses dst_ip, string dscp, string ecn, string ttl)
+{
+
+    SWSS_LOG_ENTER();
+
+    int length_of_tunnel_attrs = 5;
+
+    // adding tunnel attributes to array and writing to ASIC_DB
+    sai_attribute_t tunnel_attrs[length_of_tunnel_attrs];
+
+    // tunnel type (only ipinip for now)
+    tunnel_attrs[0].id = SAI_TUNNEL_ATTR_TYPE;
+    tunnel_attrs[0].value.s32 = SAI_TUNNEL_IPINIP;
+    tunnel_attrs[1].id = SAI_TUNNEL_ATTR_OVERLAY_INTERFACE;
+    tunnel_attrs[1].value.oid = overlayIfId;
+
+    // decap ecn mode (copy from outer/standard)
+    tunnel_attrs[2].id = SAI_TUNNEL_ATTR_DECAP_ECN_MODE;
+    if (ecn == "copy_from_outer") 
+    {
+        tunnel_attrs[2].value.s32 = SAI_TUNNEL_DECAP_ECN_MODE_COPY_FROM_OUTER;
+    }
+    else if (ecn == "standard") 
+    {
+        tunnel_attrs[2].value.s32 = SAI_TUNNEL_DECAP_ECN_MODE_STANDARD;
+    }
+
+    // ttl mode (uniform/pipe)
+    tunnel_attrs[3].id = SAI_TUNNEL_ATTR_DECAP_TTL_MODE;
+    if (ttl == "uniform") 
+    {
+        tunnel_attrs[3].value.s32 = SAI_TUNNEL_TTL_UNIFORM_MODEL;
+    }
+    else if (ttl == "pipe") 
+    {
+        tunnel_attrs[3].value.s32 = SAI_TUNNEL_TTL_PIPE_MODEL;
+    }
+
+    // dscp mode (uniform/pipe)
+    tunnel_attrs[4].id = SAI_TUNNEL_ATTR_DECAP_DSCP_MODE;
+    if (dscp == "uniform") 
+    {
+        tunnel_attrs[4].value.s32 = SAI_TUNNEL_DSCP_UNIFORM_MODEL;
+    }
+    else if (dscp == "pipe") 
+    {
+        tunnel_attrs[4].value.s32 = SAI_TUNNEL_DSCP_PIPE_MODEL;
+    }
+
+    // write attributes to ASIC_DB
+    sai_object_id_t tunnel_id;
+    sai_status_t status = sai_tunnel_api->create_tunnel(&tunnel_id, length_of_tunnel_attrs, tunnel_attrs);
```

Comment:
what's if just change one attribute of existing tunnel? are we going to create a new one? can we maintain a tunnel cache and use set_attribute to change the attribute accordingly?